### ISSUE TITLE
BGP4+ - Using link-local addresses for BGP Peering

### ISSUE THREAD

**Issue by nskeeper** [CumulusNetworks/quagga#13]
Hi,

Well, for IPv6 link state protocols (e.g. OSPFv3, IS-IS) there *is* no global next hop, thus only the link-local next hop can be shown. 
As far as I can see from my Cisco routers the situation is the same here. 

For BGP the situation is different since you can have a next hop defined by an IPv6 address (not directly connected, link-local cannot be used, e.g., ebgp.) - and then that next hop needs to be recursively resolved to a directly connected interface and IPv6 address (may be link-local or global). 

I would expect to see the next hop not the link-local. That's all. if I go on my other routers (ie Cisco) and do a show ipv6 route I get the global addressed as next hop. I expected to see the same behavior.

To sum it up a little: 
What are the pros/cons of using link.local addresses for the BGP Peering? 

Any Input or hints to documents are much appreciated. 

PS.
How can quagga like CISCO routers use global addresses as the next hop installing in the RIB?

Thanks, 
LiHongguang

**Comment by AdrianSimionov** [comment]
@nskeeper 

Q: What are the pros/cons of using link.local addresses for the BGP Peering?

A: Link-local addresses are generated and dynamic.
From a design perspective, it is never recommended to add a dynamic or generated input to a static config.

RFC 4291 defines the link local addresses. You can take a look and assume what will happen if the interface ID will change, or the MAC address will change in one of the machines.

If you will assume the link-local will never change, then you are safe, but remember that it is an assumption which can create a point of failure in the future which is not needed.

Link local is a TCP concept, that is why OSPF will not use this concept. BGP on the other hand relies on TCP for communication.

Regards,
Adrian.

**Comment by nskeeper** [comment]
Hi Adrian,

Thanks for your reply!

Could you tell me why does routers made by vendors, e.g., cisco, juniper and huawei running  bgp use the global addresses as the next hop  in the RIB?

Best Regards,

LiHongguang

**Comment by AdrianSimionov** [comment]
It is vendor specific how each of them implements the RIB internally. This is actually the reason why you choose one vendor or another.

I do not know why specifically these vendors did it this way.

Routing decisions would require additional gates on the FPGA maybe if the RIB would have the link local information???

My understanding from your original question was that Cisco displays both link local and global addresses in the RIB output and Quagga displays only the link local as next hop and you are looking for either both or only the global address?

If this is the case, I would consider your request a feature request.

Adrian.

**Comment by nskeeper** [comment]
Cisco displays global addresses in the RIB output , but Quagga displays only the link local as next hop in the RIB (installing in zebra) and the global addresses in the BGP RIB (installing in bgp).
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The network topology under test:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
R1(g0/0:1::1/64)------------------(g0/0:1::2/64)R2-----|loopback0(2::1/64)

1,The output info from cisco:
a, show bgp ipv6 unicast on R1         
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
BGP table version is 2, local router ID is 1.1.1.1
Status codes: s suppressed, d damped, h history, * valid, > best, i - internal,
              r RIB-failure, S Stale, m multipath, b backup-path, x best-external, f RT-Filter
Origin codes: i - IGP, e - EGP, ? - incomplete

   Network          Next Hop            Metric LocPrf Weight Path
*>i2::/64           1::2                     0    100      0 i
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
b,show ipv6 route on R1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
IPv6 Routing Table - default - 4 entries
Codes: C - Connected, L - Local, S - Static, U - Per-user Static route
       B - BGP, HA - Home Agent, MR - Mobile Router, R - RIP
       I1 - ISIS L1, I2 - ISIS L2, IA - ISIS interarea, IS - ISIS summary
       D - EIGRP, EX - EIGRP external, NM - NEMO, ND - Neighbor Discovery
       l - LISP
       O - OSPF Intra, OI - OSPF Inter, OE1 - OSPF ext 1, OE2 - OSPF ext 2
       ON1 - OSPF NSSA ext 1, ON2 - OSPF NSSA ext 2
C   1::/64 [0/0]
     via GigabitEthernet0/0, directly connected
L   1::1/128 [0/0]
     via GigabitEthernet0/0, receive
B   2::/64 [200/0]
     via 1::2
L   FF00::/8 [0/0]
     via Null0, receive
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2,The output info from Quagga:
a, show bgp ipv6 unicast on R1         
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
BGP table version is 0, local router ID is 2.2.2.2
Status codes: s suppressed, d damped, h history, * valid, > best, = multipath,
              i internal, r RIB-failure, S Stale, R Removed
Origin codes: i - IGP, e - EGP, ? - incomplete

   Network          Next Hop            Metric LocPrf Weight Path
*>i2::/64           1::1                     0    100      0 i

Total number of prefixes 1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
b,show ipv6 route on R1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Codes: K - kernel route, C - connected, S - static, R - RIPng,
       O - OSPFv6, I - IS-IS, B - BGP, A - Babel, T - Table,
       > - selected route, * - FIB route

C>* 1::/64 is directly connected, eth0
B>* 2::/64 [200/0] via fe80::d841:71ff:fe7f:16e6, eth0, 00:17:29

LiHongguang.

**Comment by AdrianSimionov** [comment]
My understanding was that where Quagga is running, there is only one RIB, the Kernel Routing Table and all the daemons are manipulating the kernel routing table. Am I wrong?

bgpd and zebra daemons will just display the Kernel Routing Table in a different format. 

I just want to be sure your issue is clear for everyone. What you are looking for is a cosmetic issue related to how RIB looks like in the show commands, right?

Adrian.

**Comment by nskeeper** [comment]
I think you are wrong.
Only zebra daemon manipulates the kernel routing table.

**Comment by AdrianSimionov** [comment]
No problem. But for me it is still not clear where your issue is. And if it is the same with the developers, it has no chance to be fixed.

Adrian.

**Comment by nskeeper** [comment]
I would expect to see the next hop not the link-local. That's all. if I go on my other routers (ie Cisco) and do a show ipv6 route I get the global addressed as next hop. I expected to see the same behavior.

**Comment by AdrianSimionov** [comment]
Then I think it is clear, you want a new feature as this is completely vendor independent. As CumulusNetworks is a commercial company, I think the best approach would be to contact their Sales Engineering team and prioritize your feature request compared to the ones from their customers.

Adrian.

**Comment by nskeeper** [comment]
Hi Adrian,

If i add "nexthop-local unchanged" for every neighbor, I will get what i want.

Thank you anyway.

PS.
The information from the link below can be useful to explained the problem to me.

http://docs.cumulusnetworks.com/display/DOCS/Configuring+Border+Gateway+Protocol+-+BGP
    ---->Advanced: Understanding How Next-hop Fields Are Set

LiHongguang.

**Comment by donaldsharp** [comment]
Adrian -

All routing daemons in the Quagga suite, communicate with Zebra the routes that they would like installed and zebra communicates the selected route into the kernel.  

LiHongguang -

Can you point to a specific way that routing is not actually working?  If you want the global address instead of the link local address than that is a feature request as Adrian has pointed out.

**Comment by nskeeper** [comment]
Thank you both. I got it.
